daemon/cmd/ciliumendpoints.go

```diff
@@ -0,0 +1,79 @@
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+)
+
+type localEndpointCache interface {
+	LookupPodName(name string) *endpoint.Endpoint
+}
+
+// This must only be run after K8s Pod and CES/CEP caches are synced and local endpoint restoration is complete.
+func (d *Daemon) cleanStaleCEPs(ctx context.Context, eps localEndpointCache, ciliumClient ciliumv2.CiliumV2Interface, enableCiliumEndpointSlice bool) error {
+	crdType := "ciliumendpoint"
+	if enableCiliumEndpointSlice {
+		crdType = "ciliumendpointslice"
+	}
+	indexer := d.k8sWatcher.GetIndexer(crdType)
+	objs, err := indexer.ByIndex("localNode", node.GetCiliumEndpointNodeIP())
+	if err != nil {
+		return fmt.Errorf("could not get %s objects from localNode indexer: %w", crdType, err)
+	}
+	if enableCiliumEndpointSlice {
+		for _, cesObj := range objs {
+			ces, ok := cesObj.(*cilium_v2a1.CiliumEndpointSlice)
+			if !ok {
+				return fmt.Errorf("unexpected object type returned from ciliumendpointslice store: %T", cesObj)
+			}
+			for i := range ces.Endpoints {
+				cep := ces.Endpoints[i]
+				if cep.Networking.NodeIP == node.GetCiliumEndpointNodeIP() {
+					d.checkCiliumEndpoint(ctx, ces.Namespace, cep.Name, ciliumClient, eps)
+				}
+			}
+		}
+	} else {
+		for _, cepObj := range objs {
+			cep, ok := cepObj.(*types.CiliumEndpoint)
+			if !ok {
+				return fmt.Errorf("unexpected object type returned from ciliumendpoint store: %T", cepObj)
+			}
+
+			if cep.Networking.NodeIP == node.GetCiliumEndpointNodeIP() {
+				d.checkCiliumEndpoint(ctx, cep.Namespace, cep.Name, ciliumClient, eps)
+			}
+		}
+	}
+	return nil
+}
+
+func (d *Daemon) checkCiliumEndpoint(ctx context.Context, cepNamespace, cepName string, ciliumClient ciliumv2.CiliumV2Interface, eps localEndpointCache) {
+	if eps.LookupPodName(cepNamespace+"/"+cepName) == nil {
+		// There exists a local CiliumEndpoint that is not in the endpoint manager.
+		// This function is run after completing endpoint restoration from local state and K8s cache sync.
+		// Therefore, we can delete the CiliumEndpoint as it is not referencing a Pod that is being managed.
+		// This may occur for various reasons:
+		// * Pod was restarted while Cilium was not running (likely prior to CNI conf being installed).
+		// * Local endpoint was deleted (i.e. due to reboot + temporary filesystem) and Cilium or the Pod where restarted.
+		log.WithFields(logrus.Fields{logfields.CEPName: cepName, logfields.K8sNamespace: cepNamespace}).
+			Info("Found stale ciliumendpoint for local pod that is not being managed, deleting.")
+		if err := ciliumClient.CiliumEndpoints(cepNamespace).Delete(ctx, cepName, metav1.DeleteOptions{}); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{logfields.CEPName: cepName, logfields.K8sNamespace: cepNamespace}).
+				Error("Could not clean stale CEP")
```

Comment:
What if there is a CEP with the same name created on a different node?

Reply:
If the CEP was created on a different Node but is referencing this node with NodeIP then its still stale and it will and should be deleted.

However, I think a better question is what if a stale CEP is deleted and recreated by a different Node between the check and delete. It's possible this could delete a valid CEP (similar to the EP sync race condition problem).

I think the solution is to add the UID to the delete precondition. That's easy enough for the CEP, but we don't store CES CEP UIDs in the CES store, which makes this a bit tricky.

Reply:
@aanm  The never ending saga of CEP ownership issues, I think the last commit should adequately address this issue at the very rare cost of an extra API call when CES is enabled:

https://github.com/cilium/cilium/pull/20350/commits/621a1004914ff743d1b9f3b80e0deca3870de6fc